Add IgnoreMap and IgnoreList options to Map List Config

PhasecoreX opened the pull request:
I have a `VoteSysMapList.ini` like this:
```
[CTF_Maps]
IncludeMapsWithPrefix=CTF-
IgnoreList=CTF_XV_Maps
IgnoreList=CTF_W00t_Maps
IgnoreMap=CTF-Beatitude

[CTF_XV_Maps]
IncludeMapsWithPrefix=CTF-XV-

[CTF_W00t_Maps]
Map=CTF-Airie
Map=CTF-AllYourFacingW00t
Map=CTF-AllYourW00t-SE
Map=CTF-AnotherBrickInTheW00t
```
With this pull request, the `LoadMapList()` function is recursively called to load each `IgnoreList` into one `VS_MapList`, which is then used to filter out maps from any `IncludeMapsWithPrefix` entries. It also allows `IgnoreMap` for ignoring one-off maps, and it won't filter out maps included with `Map`.

Add `IncludeMapsWithPrefix` -> filter out `IgnoreList` and `IgnoreMap` -> add back `Map`. (The code doesn't actually work in that order, but logically it does).

I tried to keep this as simple as possible, and in the same style. Please feel free to make any edits, or let me know if you want me to make any changes. The only thing that sort of bugs me is that when it recursively loads `IgnoreList` entries, it shows the `Loading List 'CTF_XV_Maps' for Botpack.CTFGame` log message, which is technically true, but you can't tell if it's loading for including or excluding. But I thought adding another parameter to the `LoadMapList()` just to change a log statement was overkill.

PhasecoreX committed:
Add IgnoreMap and IgnoreList options to Map List Config

Deaod commented:
I am not a fan of only applying ignored maps to `IncludeMapsWithPrefix`.

I thought about it a bit and i think you could create the same effect by first adding all maps from `Map` and `IncludeMapsWithPrefix`, then subtracting `IgnoreMap` and `IgnoreList`.  
In addition to this, a new `IncludeList` member should be added, which also adds all of its maps to the list before ignored maps are removed.

If you wanted to achieve the same effect as what you implemented, youd have to have two levels of map lists like so:
```ini
[CTF_Maps_With1w00t]
IncludeList=CTF_Maps
Map=CTF-AllYourW00t-SE

[CTF_Maps]
IncludeMapsWithPrefix=CTF-
IgnoreList=CTF_XV_Maps
IgnoreList=CTF_W00t_Maps
IgnoreMap=CTF-Beatitude

[CTF_XV_Maps]
IncludeMapsWithPrefix=CTF-XV-

[CTF_W00t_Maps]
Map=CTF-Airie
Map=CTF-AllYourFacingW00t
Map=CTF-AllYourW00t-SE
Map=CTF-AnotherBrickInTheW00t
```

The logic to me seems simpler. In List1 you first add all maps from positive members, then you subtract maps through negative members. What you get is a non-empty list of maps. If you need to add something back in because you subtracted too much, you create a new list List2 that contains includes all maps of List1, *plus* some maps that might have previously been subtracted.

Does this make sense to you?

P.S.: Please rebase your changes to hopefully fix the CI that i broke.

PhasecoreX committed:
Merge branch 'Deaod:main' into main

Deaod commented:
I am not in principle against merging this, but I wanted your opinion on `IncludeList` and confirm with you that you deliberately chose the way the other settings interact.

If you still think the way its arranged in your PR is correct, im happy to work from there to add `IncludeList`, unless you also want to try your hand at adding that.

One more thing though i would like to ask of you: Please add a bit of documentation to README.md and an example to VoteSysMapLists.ini. Please make sure to document how the different settings interact, because i fear it wont be obvious to admins that dont already know how it works.

PhasecoreX commented:
I will work on adding the above once I am done with work for the day, no worries. I will have it so that all ignores will override any included maps, like you described above. Additionally, I will have `IncludeList` as well as a `IgnoreMapsWithPrefix`, so that we have 3 ways of including maps as well as the same 3 ways of ignoring them.
```
Map
IgnoreMap

IncludeMapsWithPrefix
IgnoreMapsWithPrefix

IncludeList
IgnoreList
```
Also going to clean up and refactor some stuff, as there is duplicated code now.

PhasecoreX commented:
I will also update the documentation!

PhasecoreX committed:
Add IncludeList and IgnoreMapsWithPrefix

Refactor map list modifying functions
Update readme

PhasecoreX commented:
Okay, so now there's the above 6 fields for adding and removing maps from a map list. I've refactored it since it would have been a bunch of duplicated code. Documentation is also updated, as well as the example .ini file.

I had initially opted to add back `Map` after the ignores because it was the only way to re-add ignored maps. But then you showed a way cleaner method of just making a new map list that makes much more sense.

If you need me to update the documentation more, let me know. I tried to make it as complete but succinct as possible.

Deaod commented on Classes/MutVoteSys.uc:
Maybe pull the logic for
```unrealscript
MapName = MapArray[i];
if (Right(MapName, 4) ~= ".unr")
	MapName = Left(MapName, Len(MapName) - 4); // we dont care about extension
```
into its own function like `CleanMapName()`.

Deaod commented on Classes/MutVoteSys.uc:
```diff
@@ -1137,28 +1138,20 @@ function VS_Map LoadMapList(class<GameInfo> Game, name ListName) {
-		for (i = 0; i < MC.IncludeMapsWithPrefix.Length; i++) {
-			if (MC.IncludeMapsWithPrefix[i] == "")
-				continue;
+		IgnoreList = MLIgnore.DuplicateList();
```
```suggestion
		IgnoreList = MLIgnore.First;
```

Deaod commented on Classes/MutVoteSys.uc:
```diff
@@ -1192,6 +1185,80 @@ function VS_Map LoadMapList(class<GameInfo> Game, name ListName) {
+		if (MapListArray[i] == '')
+			continue;
+
+		IncludeList = LoadMapList(Game, MapListArray[i]);
```
This looks dangerous, because if `MapListArray[i]` is an empty name or `'None'`, all maps for Game will be included. This is fine when iterating through `IncludeList`, but not so much when iterating through `IgnoreList`.

I think LoadMapList should return `none` immediately if Game is `none` and ListName is empty.

You also shouldnt have to pass Game through all of these then and can just pass `none` to `LoadMapList()` then.

Deaod reviewed (changes requested):
I like the documentation. I like the general idea of these changes. I really like that youve done this work.

If youre willing to do a few more changes, please look at the comments ive left. If not, let me know and ill merge this and fix it myself.

Deaod commented:
A future topic in this direction will be reducing the complexity of building the lists. The number of comparisons for n potential maps added to a list with m maps excluded is n x m, which is sure to break the op-count limit eventually.

I do not mean this as something you need to fix immediately, but ill add this as an issue for the future.

PhasecoreX committed:
Merge branch 'Deaod:main' into main

PhasecoreX commented:
I will make the changes. Yeah I couldn't think of a better way of doing "if string in list of string" besides the O(n^2) loop inside loop thing. A sort of HashSet type thing would be good, but I don't think UScript has that, it would have to be made manually.

PhasecoreX committed:
Add CleanMapName() function

Update AddMapListsToMapList to not use Game

Deaod commented:
Thank you very much. These are really nice changes.
Do you have a specific name you want to be credited as?

PhasecoreX commented:
You're welcome! PhasecoreX is fine!
